r/tools/nixlibs.R

```diff
@@ -473,17 +473,20 @@ build_libarrow <- function(src_dir, dst_dir) {
   env_vars <- env_vars_as_string(env_var_list)
 
   cat("**** arrow", ifelse(quietly, "", paste("with", env_vars)), "\n")
-  status <- suppressWarnings(system(
-    paste(env_vars, "inst/build_arrow_static.sh"),
-    ignore.stdout = quietly, ignore.stderr = quietly
+
+  status <- suppressWarnings(system2(
+    "bash",
```

Comment:
I'm not sure you need this, we didn't before.

Reply:
+1 on using a file and `cat`ing that in full if build fails. With the nature of CMake errors it is not always the case that the error is in the last lines (due to parallelism) and the issue with non-portable flag false positives is not an issue this way as the build has already failed...

Reply:
My experience with build failures matches what @assignUser refers to with having to scroll up (sometimes quite far) to find the source of the failure. One trick I use to bring the error down is to re-build without parallelism but I'm not sure how best to implement that here.

So I'd be in favor of (1) writing to a file instead of capturing and (2) cating the full file to hedge against the tempfile disappearing before someone can grab it. I'll do that now.

Reply:
@nealrichardson, Re:

> I'm not sure you need this, we didn't before.

Were you referring to how I broke up the args in the system2 call? Calling the script the old way does work with `system` but we switched to `system2 `to capture stdin/stdout and calling system2 the same way resulted in this error,

> sh: SOURCE_DIR='../cpp' BUILD_DIR='/var/folders/vm/bbcxtl3s63l7gn9h1tvfh1mm0000gn/T//RtmpoTYnF4/file166a06be38857' DEST_DIR='libarrow/arrow-13.0.0.9000' CMAKE='/opt/homebrew/bin/cmake' EXTRA_CMAKE_FLAGS='' CC='clang -arch arm64' CXX='clang++ -arch arm64 -std=gnu++17' LDFLAGS='-L/opt/R/arm64/lib' N_JOBS='2' ARROW_S3='OFF' ARROW_GCS='OFF' inst/build_arrow_static.sh: No such file or directory

I think this is because of what the man page for `system2` says, which is:

> Unlike [system](http://127.0.0.1:17041/help/library/base/help/system), command is always quoted by [shQuote](http://127.0.0.1:17041/help/library/base/help/shQuote), so it must be a single command without arguments.

So I felt like my change to the args made sense.

Reply:
Alright, with https://github.com/apache/arrow/pull/37727/commits/4246503793ca46040c4b941469a5876268f7c965, we now redirect stdout/stderr to a tempfile and print that file out on a build failure. The way I set it up, it looks like

> **** Error building Arrow C++.
> **** Printing contents of build log because the build failed while ARROW_R_DEV was set to FALSE
> /* contents of build log */
> **** Complete build log is available at /var/folders/vm/bbcxtl3s63l7gn9h1tvfh1mm0000gn/T//RtmpKBAT6R/file75d2a7f76ba.log

I printed the build log last in case we print so much information in the prior step that the user's terminal truncates the first part of the output so the user can still find the file (if it exists at this point).

Reply:
I was referring to the next line, looks like the mobile app attached the comment to the wrong line. I'm not sure that directly calling "bash" is necessary or a good idea--we didn't before, the script has a shebang at the top, and bash might not be the shell that's available, a point that CRAN has been pedantic about in the past.

Reply:
Gotcha, that makes sense now. I thought the pre-existing reliance on bash in nixlibs.R was odd, and we actually have this bit:

https://github.com/apache/arrow/blob/d9ee3a7961df44a1e30d3fe615ee3a20af569025/r/tools/nixlibs.R#L625-L627

and another spot where we make the `command` bash,

https://github.com/apache/arrow/blob/d9ee3a7961df44a1e30d3fe615ee3a20af569025/r/tools/nixlibs.R#L630

So this change adds some consistency. I'd be happy to changes things if/when CRAN has any issues.